### PR TITLE
fix: SessionExpired issue

### LINKED ISSUE #987: Session expired while scanning aws
**Bug Template**

*Title*: *neo4j.exceptions.SessionExpired: Failed to read from defunct connection IPv4Address(('07be756b.databases.neo4j.io', 7687)) *

*Description*:
running cartography in ec2 launched instance to observe relationship of my aws resources

*To Reproduce*:
> cartography --neo4j-uri neo4j+s://07be756b.databases.neo4j.io --neo4j-password-prompt --neo4j-user neo4j

*Logs*:
`Unhandled exception during sync stage 'aws'Traceback (most recent call last):  File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 57, in _yield_messages    buffer.receive(sock, 2)  File "/usr/local/lib/python3.7/site-packages/neo4j/packstream.py", line 484, in receive    raise OSError("No data")OSError: No dataThe above exception was the direct cause of the following exception:Traceback (most recent call last):  File "/home/ec2-user/cartography-ec2-dev/cartography/sync.py", line 86, in run    stage_func(neo4j_session, config)  File "/home/ec2-user/cartography-ec2-dev/cartography/util.py", line 132, in timed    return method(*args, **kwargs)  File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/__init__.py", line 244, in start_aws_ingestion    requested_syncs,  File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/__init__.py", line 168, in _sync_multiple_accounts    aws_requested_syncs=aws_requested_syncs,  # Could be replaced later with per-account requested syncs  File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/__init__.py", line 63, in _sync_one_account    RESOURCE_FUNCTIONS[func_name](**sync_args)  File "/home/ec2-user/cartography-ec2-dev/cartography/util.py", line 132, in timed    return method(*args, **kwargs)  File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/ec2/volumes.py", line 118, in sync_ebs_volumes    load_volume_relationships(neo4j_session, transformed_data, update_tag)  File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/ec2/volumes.py", line 95, in load_volume_relationships    aws_update_tag=aws_update_tag,  File "/home/ec2-user/cartography-ec2-dev/cartography/experimental_neo4j_4x_support.py", line 109, in wrapper    return run_func(query, *args, **kwargs)  File "/usr/local/lib/python3.7/site-packages/neo4j/work/simple.py", line 218, in run    self._bookmarks, **kwparameters  File "/usr/local/lib/python3.7/site-packages/neo4j/work/result.py", line 117, in _run    self._attach()  File "/usr/local/lib/python3.7/site-packages/neo4j/work/result.py", line 206, in _attach    self._connection.fetch_message()  File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 188, in inner    func(*args, **kwargs)  File "/usr/local/lib/python3.7/site-packages/neo4j/io/_bolt4.py", line 283, in fetch_message    details, summary_signature, summary_metadata = next(self.inbox)  File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 83, in __next__    tag, fields = self.pop()  File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 77, in pop    return next(self._messages)  File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 74, in _yield_messages    self.on_error(error)  File "/usr/local/lib/python3.7/site-packages/neo4j/io/__init__.py", line 561, in _set_defunct_read    self._set_defunct(message, error=error, silent=silent)  File "/usr/local/lib/python3.7/site-packages/neo4j/io/__init__.py", line 606, in _set_defunct    raise SessionExpired(message) from errorneo4j.exceptions.SessionExpired: Failed to read from defunct connection IPv4Address(('07be756b.databases.neo4j.io', 7687)) (IPv4Address(('34.126.171.25', 7687))) | Unhandled exception during sync stage 'aws' Traceback (most recent call last): File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 57, in _yield_messages buffer.receive(sock, 2) File "/usr/local/lib/python3.7/site-packages/neo4j/packstream.py", line 484, in receive raise OSError("No data") OSError: No data The above exception was the direct cause of the following exception: Traceback (most recent call last): File "/home/ec2-user/cartography-ec2-dev/cartography/sync.py", line 86, in run stage_func(neo4j_session, config) File "/home/ec2-user/cartography-ec2-dev/cartography/util.py", line 132, in timed return method(*args, **kwargs) File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/__init__.py", line 244, in start_aws_ingestion requested_syncs, File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/__init__.py", line 168, in _sync_multiple_accounts aws_requested_syncs=aws_requested_syncs, # Could be replaced later with per-account requested syncs File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/__init__.py", line 63, in _sync_one_account RESOURCE_FUNCTIONS[func_name](**sync_args) File "/home/ec2-user/cartography-ec2-dev/cartography/util.py", line 132, in timed return method(*args, **kwargs) File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/ec2/volumes.py", line 118, in sync_ebs_volumes load_volume_relationships(neo4j_session, transformed_data, update_tag) File "/home/ec2-user/cartography-ec2-dev/cartography/intel/aws/ec2/volumes.py", line 95, in load_volume_relationships aws_update_tag=aws_update_tag, File "/home/ec2-user/cartography-ec2-dev/cartography/experimental_neo4j_4x_support.py", line 109, in wrapper return run_func(query, *args, **kwargs) File "/usr/local/lib/python3.7/site-packages/neo4j/work/simple.py", line 218, in run self._bookmarks, **kwparameters File "/usr/local/lib/python3.7/site-packages/neo4j/work/result.py", line 117, in _run self._attach() File "/usr/local/lib/python3.7/site-packages/neo4j/work/result.py", line 206, in _attach self._connection.fetch_message() File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 188, in inner func(*args, **kwargs) File "/usr/local/lib/python3.7/site-packages/neo4j/io/_bolt4.py", line 283, in fetch_message details, summary_signature, summary_metadata = next(self.inbox) File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 83, in __next__ tag, fields = self.pop() File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 77, in pop return next(self._messages) File "/usr/local/lib/python3.7/site-packages/neo4j/io/_common.py", line 74, in _yield_messages self.on_error(error) File "/usr/local/lib/python3.7/site-packages/neo4j/io/__init__.py", line 561, in _set_defunct_read self._set_defunct(message, error=error, silent=silent) File "/usr/local/lib/python3.7/site-packages/neo4j/io/__init__.py", line 606, in _set_defunct raise SessionExpired(message) from error neo4j.exceptions.SessionExpired: Failed to read from defunct connection IPv4Address(('07be756b.databases.neo4j.io', 7687)) (IPv4Address(('34.126.171.25', 7687)))`

*Screenshots*:
> If applicable, add screenshots to help explain your problem.

*Please complete the following information:*:
>  - Cartography release version or commit hash [1e274e3e1e30e6558a688d70f4f274b5e78f8dc0]
>  - Python version: 2.7.18
>  - OS (EC2-Amazon Linux 2)

*Additional context*:
> Add any other context about the problem here.

### PR DESCRIPTION
**Pull request by AlmogBaku**
Introducing a SessionWrapper that is responsible for renewing the Neo4J driver and session in case of SessionExpired exception.

fixes #987

### CODE REVIEW

**Comment by ramonpetgrave64** [comment]
Hey! Thanks for this. This could be a good solution, until the neo4j client libraries become more flexible with timeouts. Can you into potential performance impacts?

**Comment by achantavy** [comment]
Hi @AlmogBaku, thank you for putting this together. Couple of questions:

1. Can you provide an example stack trace of experiencing #987 with cartography? Asking because I am wondering if this only occurs in cases where `session.run()` is called, and I would be very interested to see a stack trace where it fails with `session.write_transaction`, because `write_transaction`'s automatic retries have solved similar-looking problems for me.

2. The Neo4j docs you [linked](https://github.com/lyft/cartography/issues/987#issuecomment-1365270272) say that max-connection-lifetime should be configured to avoid this, and cartography does expose this setting as mentioned here: https://github.com/lyft/cartography/issues/987#issuecomment-1307985045. Has playing with this setting helped with #987 at all?

2. Do you have a console log trace showing that this code change fixes the problem in #987?

**Comment by AlmogBaku** [comment]
@ramonpetgrave64 - the performance impact is relatively minor since I only keep one instance and invalidate it upon failure.

@achantavy -
1. My stack is Neo4j Aura (GCP) + k8s job for cartography(on EKS). Apparently, the AWS gateway is dropping the connection after a while of idling. Neo4j supports claims it also happens in other use-cases.
From my research, I know it also happens with other hosted neo4j solutions.

I don't know about the `session.write_transaction`, but I guess it's the same because the driver is responsible for the TCP connection and it gets dropped.

2. Changing the timeout doesn't help because the gateway might drop it. Neo4j support team also stated that putting up a "retrying mechanism" such as this is their recommendation

3. I can't share the login trace because it also shares some sensitive data, but I tried this patch in production, and it works :) I'd be happy to provide more information via Slack DM

**Comment by achantavy** [comment]
> My stack is Neo4j Aura

Ah, I meant can you provide a log trace showing where the crash happens? This way we might be able to identify if the issue is more likely to show up in certain parts of the code.

> I don't know about the session.write_transaction, but I guess it's the same because the driver is responsible for the TCP connection and it gets dropped.

`session.write_transaction` does automatic retries when the TCP connection gets dropped and `session.run` does not. cartography has a lot of older code that still calls `session.run` that we are trying to refactor, so if we know that only `session.run` is problematic, then we can solve this issue without adding additional complexity by maintaining our own driver reisntantiation logic.

> I can't share the login trace because it also shares some sensitive data

Can you redact out the sensitive parts?

It would be helpful to see a failing log trace where the issue happens, and a successful log trace to show that the issue is solved.

Just for additional reference, we implemented max-connection-lifetime here: https://github.com/lyft/cartography/pull/535 and it helped solve timeouts with our NLBs. If you haven't already, I'd definitely recommend setting it to something less than 350s if you know that the connection is going over an AWS NLB (I don't know what Aura uses for hosting). I'd also be very interested in seeing the session expired crash even with this setting.

To summarize, the logic table I'd want to know is

|   | set --neo4j-max-connection-lifetime to less than 350 seconds |  
|---|---|
|Did it crash? (yes/no)   |   |
| If yes, provide log trace of the crash without confidential information |   |  

Transient crashes are frustrating and I definitely appreciate your help; I'm asking all these questions to gather data before deciding to own additional complexity (imo this kind of thing really should be handled by the neo4j driver :p)

**Comment by AlmogBaku** [comment]
1. Yes, I tried to reduce the conn lifetime, same error
2. The error log is **very** similar to #987. If still needed - I'll be able to reproduce it next week and send you the full log
3. I totally agree - it would have been better if the driver had handled it, but it seems like Neo4j team disagree 😖

**Comment by achantavy** [comment]
Hi @AlmogBaku, please send the log over, or if that is not possible then please reply back with what exact cartography function was running when it crashed (e.g. was it running load() in EC2 instances, or load() in IAM, or...?).

It'd also be helpful to know if it always crashes in that same spot or if there are multiple places (if you have multiple crash dumps, then great, send them _all_ if you can!).

I'd really like to know you're seeing this crash as a result of session.run() or write_transaction().

**Comment by AlmogBaku** [comment]
@achantavy see #987 for the log file

**Comment by achantavy** [comment]
Thanks @AlmogBaku for sharing the log, this is helpful. I'd like a bit of time to play with this a bit before merging. Thanks for your paitience. cc: @ramonpetgrave64 @mpurusottamc FYI.

@harshagw just wondering, how difficult would it be to combine the neo4j session factory that you wrote with the session wrapper shown in this PR?

**Comment by AlmogBaku** [comment]
any update?
